core/embed/prodtest/optiga_prodtest.c

```diff
@@ -540,33 +552,124 @@ static const uint8_t ECDSA_WITH_SHA256[] = {
 };
 // clang-format on
 
-static const uint8_t ROOT_PUBLIC_KEYS[][65] = {
-    {
-        // Production root public key.
-        0x04, 0xca, 0x97, 0x48, 0x0a, 0xc0, 0xd7, 0xb1, 0xe6, 0xef, 0xaf,
-        0xe5, 0x18, 0xcd, 0x43, 0x3c, 0xec, 0x2b, 0xf8, 0xab, 0x98, 0x22,
-        0xd7, 0x6e, 0xaf, 0xd3, 0x43, 0x63, 0xb5, 0x5d, 0x63, 0xe6, 0x03,
-        0x80, 0xbf, 0xf2, 0x0a, 0xcc, 0x75, 0xcd, 0xe0, 0x3c, 0xff, 0xcb,
-        0x50, 0xab, 0x6f, 0x8c, 0xe7, 0x0c, 0x87, 0x8e, 0x37, 0xeb, 0xc5,
-        0x8f, 0xf7, 0xcc, 0xa0, 0xa8, 0x3b, 0x16, 0xb1, 0x5f, 0xa5,
-    },
-    {
-        // Development root public key.
-        0x04, 0x7f, 0x77, 0x36, 0x8d, 0xea, 0x2d, 0x4d, 0x61, 0xe9, 0x89,
-        0xf4, 0x74, 0xa5, 0x67, 0x23, 0xc3, 0x21, 0x2d, 0xac, 0xf8, 0xa8,
-        0x08, 0xd8, 0x79, 0x55, 0x95, 0xef, 0x38, 0x44, 0x14, 0x27, 0xc4,
-        0x38, 0x9b, 0xc4, 0x54, 0xf0, 0x20, 0x89, 0xd7, 0xf0, 0x8b, 0x87,
-        0x30, 0x05, 0xe4, 0xc2, 0x8d, 0x43, 0x24, 0x68, 0x99, 0x78, 0x71,
-        0xc0, 0xbf, 0x28, 0x6f, 0xd3, 0x86, 0x1e, 0x21, 0xe9, 0x6a,
-    },
-};
+static bool get_authority_key_digest(DER_ITEM *tbs_cert,
+                                     const uint8_t **authority_key_digest) {
+  // Find the certificate extensions in the tbsCertificate.
+  DER_ITEM cert_item = {0};
+  bool found = false;
+  while (der_read_item(&tbs_cert->buf, &cert_item)) {
+    if (cert_item.id == DER_X509_EXTENSIONS) {
+      found = true;
+      break;
+    }
+  }
+
+  if (!found) {
+    vcp_println("ERROR get_authority_key_digest, extensions not found.");
+    return false;
+  }
+
+  // Open the extensions sequence.
+  DER_ITEM extensions = {0};
+  if (!der_read_item(&cert_item.buf, &extensions) ||
+      extensions.id != DER_SEQUENCE) {
+    vcp_println("ERROR get_authority_key_digest, der_read_item extensions.");
+    return false;
+  }
+
+  // Find the authority key identifier extension.
+  DER_ITEM extension = {0};
+  found = false;
+  while (der_read_item(&extensions.buf, &extension)) {
+    DER_ITEM extension_id = {0};
+    if (der_read_item(&extension.buf, &extension_id) &&
+        extension_id.buf.size == sizeof(OID_AUTHORITY_KEY_IDENTIFIER) &&
+        memcmp(extension_id.buf.data, OID_AUTHORITY_KEY_IDENTIFIER,
+               sizeof(OID_AUTHORITY_KEY_IDENTIFIER)) == 0) {
+      found = true;
+      break;
+    }
+  }
+
+  if (!found) {
+    vcp_println(
+        "ERROR get_authority_key_digest, authority key identifier extension "
+        "not found.");
+    return false;
+  }
+
+  // Find the authority key identifier extension's extnValue.
+  // Conforming CAs must mark this extension as non-critical, so there shouldn't
+  // be anything between the extension ID and value, but we search for the octet
+  // string to be on the safe side.
```

Comment:
I wrote the code before I looked up the requirement about the extension having to be non-critical. My motivation for leaving it this way was if in time we refactor the code to work with other extensions, we can use it right away instead of having to think about this again. The difference is basically getting rid of the lines where `found` appears. I don't have a problem removing it though.

Reply:
I also considered creating a general set of functions for handling certificates and putting it in `crypto`. For example `x509_get_extensions(cert, extentions)` and `x509_get_extension_value(oid, value)`, but it seems like over-engineering at the moment.

Reply:
Here is a small refactor implementing `get_cert_extensions()` and `get_extension_value()`:
https://github.com/trezor/trezor-firmware/pull/3451/commits/a67cbcffc38d0607b99b5a0f23c39e88bcad5980.